src/CodeGen_LLVM.cpp

```diff
@@ -4767,27 +4822,45 @@ Value *CodeGen_LLVM::call_overloaded_intrin(const Type &result_type, const std::
         }
 
         if (!resolved) {
-            resolved = &i;
+            debug(debug_level) << "Resolved!\n";
+            resolved = &overload;
         } else {
             if (resolved->result_type.lanes() < result_type.lanes()) {
                 // The current match is smaller than the result type. Take the bigger intrinsic.
-                if (i.result_type.lanes() > resolved->result_type.lanes()) {
-                    resolved = &i;
+                if (overload.result_type.lanes() > resolved->result_type.lanes()) {
+                    debug(debug_level) << "Replaced with bigger intrinsic\n";
+                    resolved = &overload;
                 }
             } else {
                 // The current match is bigger than the result type. If the current candidate is also bigger,
                 // but smaller than the current match, take it instead.
-                if (i.result_type.lanes() >= result_type.lanes() && i.result_type.lanes() < resolved->result_type.lanes()) {
-                    resolved = &i;
+                if (overload.result_type.lanes() >= result_type.lanes() && overload.result_type.lanes() < resolved->result_type.lanes()) {
+                    debug(debug_level) << "Replaced with smaller intrinsic\n";
+                    resolved = &overload;
                 }
             }
         }
     }
 
     if (resolved) {
-        return call_intrin(result_type, resolved->result_type.lanes(), resolved->impl, args);
+        std::vector<Expr> promoted_args;
+        promoted_args.reserve(args.size());
+        for (size_t i = 0; i < args.size(); i++) {
+            Expr promoted_arg = args[i];
+            if (args[i].type().is_scalar()) {
+                promoted_arg = lossless_cast(resolved->arg_types[i].element_of(), promoted_arg);
+            }
+            if (resolved->arg_types[i].is_vector() && args[i].type().is_scalar() && result_type.lanes() > 1) {
+                // We're passing a scalar to a vector argument, broadcast it.
+                promoted_args.emplace_back(Broadcast::make(promoted_arg, result_type.lanes()));
```

Comment:
Is there any point in using emplace_back for an rvalue of the vector element type? Doesn't push_back have an overload that takes an rvalue reference and std::moves it into the vector? I'm only been using emplace_back when I want to pass constructor args for the object I'm pushing, and now I'm wondering if I've been doing it wrong.

Reply:
I think emplace_back is a superset of push_back, so just using it everywhere seems reasonable? I got some cases of clang-tidy complaining about push_back as well (probably not this one though).